src/net/socket_utils.zig

```diff
@@ -10,125 +9,109 @@ const Channel = sig.sync.Channel;
 const Logger = sig.trace.Logger;
 const ExitCondition = sig.sync.ExitCondition;
 
-const UdpSocket = @import("zig-network").Socket;
+const network = @import("zig-network");
+const UdpSocket = network.Socket;
 
 pub const SOCKET_TIMEOUT_US: usize = 1 * std.time.us_per_s;
 pub const PACKETS_PER_BATCH: usize = 64;
 
 // The identifier for the scoped logger used in this file.
 const LOG_SCOPE: []const u8 = "socket_utils";
 
-pub fn readSocket(
-    socket_: UdpSocket,
-    incoming_channel: *Channel(Packet),
-    logger_: Logger,
-    exit: ExitCondition,
-) !void {
-    const logger = logger_.withScope(LOG_SCOPE);
-    defer {
-        exit.afterExit();
-        logger.info().log("readSocket loop closed");
-    }
-
-    // NOTE: we set to non-blocking to periodically check if we should exit
-    var socket = socket_;
-    try socket.setReadTimeout(SOCKET_TIMEOUT_US);
-
-    while (exit.shouldRun()) {
-        var packet: Packet = Packet.default();
-        const recv_meta = socket.receiveFrom(&packet.data) catch |err| switch (err) {
-            error.WouldBlock => continue,
-            else => |e| {
-                logger.err().logf("readSocket error: {s}", .{@errorName(e)});
-                return e;
-            },
-        };
-        const bytes_read = recv_meta.numberOfBytes;
-        if (bytes_read == 0) return error.SocketClosed;
-        packet.addr = recv_meta.sender;
-        packet.size = bytes_read;
-        try incoming_channel.send(packet);
-    }
-}
-
-pub fn sendSocket(
-    socket: UdpSocket,
-    outgoing_channel: *Channel(Packet),
-    logger_: Logger,
-    exit: ExitCondition,
-) !void {
-    const logger = logger_.withScope(LOG_SCOPE);
-    defer {
-        // empty the channel
-        while (outgoing_channel.tryReceive()) |_| {}
-        exit.afterExit();
-        logger.debug().log("sendSocket loop closed");
-    }
-
-    while (exit.shouldRun()) {
-        while (outgoing_channel.tryReceive()) |p| {
-            const bytes_sent = socket.sendTo(p.addr, p.data[0..p.size]) catch |e| {
-                logger.err().logf("sendSocket error: {s}", .{@errorName(e)});
-                continue;
-            };
-            std.debug.assert(bytes_sent == p.size);
-        }
-    }
-}
-
-/// A thread that is dedicated to either sending or receiving data over a socket.
-/// The included channel can be used communicate with that thread.
-///
-/// The channel only supports one: either sending or receiving, depending how it
-/// was initialized. While you *could* send data to the channel for a "receiver"
-/// socket, the underlying thread won't actually read the data from the channel.
-pub const SocketThread = struct {
-    channel: *Channel(Packet),
-    exit: *Atomic(bool),
+pub const SocketPipe = struct {
     handle: std.Thread,
 
     const Self = @This();
 
-    pub fn initSender(
+    pub const Direction = enum {
+        sender,
+        receiver,
+    };
+
+    pub fn init(
```

Comment:
do you think it would be better to have two seperate methods initSender and initReciever ? -- direction seems almost hidden with how key it is to how this method operates

Reply:
https://github.com/Syndica/sig/pull/486/commits/767607c3c46beb88cf4533d9bbabba71b9b3e634